File: backend/app/schemas/simulation.py
from pydantic import BaseModel, Field, field_validator


class SimulationRequest(BaseModel):
    scenario_name: str = Field(min_length=1, description="Scenario name for traceability.")
    population_size: int = Field(ge=10, le=10000, description="Synthetic population size.")
    cycles: int = Field(ge=1, le=100, description="Number of simulation cycles.")

    @field_validator("scenario_name")
    @classmethod
    def validate_scenario_name(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("scenario_name must not be empty.")
        return normalized


class SimulationCycleResult(BaseModel):
    cycle: int = Field(ge=1)
    supporters: int = Field(ge=0)
    resistant: int = Field(ge=0)
    hesitant: int = Field(ge=0)


class SimulationFinalResult(BaseModel):
    supporters: int = Field(ge=0)
    resistant: int = Field(ge=0)
    hesitant: int = Field(ge=0)
    supporters_percentage: float = Field(ge=0, le=100)
    resistant_percentage: float = Field(ge=0, le=100)
    hesitant_percentage: float = Field(ge=0, le=100)


class SimulationInterpretation(BaseModel):
    dominant_trend: str
    social_risk: str
    polarization_level: str


class SimulationScenarioSnapshot(BaseModel):
    final_result: SimulationFinalResult
    timeline: list[SimulationCycleResult]
    interpretation: SimulationInterpretation


class SimulationResponse(BaseModel):
    scenario_name: str
    population_size: int
    cycles: int
    final_result: SimulationFinalResult
    timeline: list[SimulationCycleResult]
    interpretation: SimulationInterpretation


class PolicyInput(BaseModel):
    institutional_trust_modifier: float = Field(ge=-1.0, le=1.0)
    media_influence: float = Field(ge=0.0, le=1.0)
    religious_influence: float = Field(ge=0.0, le=1.0)
    economic_stress: float = Field(ge=0.0, le=1.0)
    policy_intensity: float = Field(ge=0.0, le=1.0)


class PolicySimulationRequest(BaseModel):
    scenario_name: str = Field(min_length=1, description="Scenario name for traceability.")
    population_size: int = Field(ge=10, le=10000, description="Synthetic population size.")
    cycles: int = Field(ge=1, le=100, description="Number of simulation cycles.")
    policy: PolicyInput

    @field_validator("scenario_name")
    @classmethod
    def validate_scenario_name_policy(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("scenario_name must not be empty.")
        return normalized


class PolicyDifference(BaseModel):
    supporters_change: int
    supporters_percentage_change: float
    resistant_change: int
    resistant_percentage_change: float
    hesitant_change: int
    hesitant_percentage_change: float
    risk_change: str
    polarization_change: str


class PolicySimulationResponse(BaseModel):
    scenario_name: str
    population_size: int
    cycles: int
    policy: PolicyInput
    base_scenario: SimulationScenarioSnapshot
    policy_scenario: SimulationScenarioSnapshot
    difference: PolicyDifference

File: frontend/src/services/simulation.ts
export type SimulationRequest = {
  scenario_name: string;
  population_size: number;
  cycles: number;
};

export type SimulationCycleResult = {
  cycle: number;
  supporters: number;
  resistant: number;
  hesitant: number;
};

export type SimulationFinalResult = {
  supporters: number;
  resistant: number;
  hesitant: number;
  supporters_percentage: number;
  resistant_percentage: number;
  hesitant_percentage: number;
};

export type SimulationInterpretation = {
  dominant_trend: string;
  social_risk: string;
  polarization_level: string;
};

export type SimulationScenarioSnapshot = {
  final_result: SimulationFinalResult;
  timeline: SimulationCycleResult[];
  interpretation: SimulationInterpretation;
};

export type SimulationResponse = {
  scenario_name: string;
  population_size: number;
  cycles: number;
  final_result: SimulationFinalResult;
  timeline: SimulationCycleResult[];
  interpretation: SimulationInterpretation;
};

export type PolicyInput = {
  institutional_trust_modifier: number;
  media_influence: number;
  religious_influence: number;
  economic_stress: number;
  policy_intensity: number;
};

export type PolicySimulationRequest = {
  scenario_name: string;
  population_size: number;
  cycles: number;
  policy: PolicyInput;
};

export type PolicyDifference = {
  supporters_change: number;
  supporters_percentage_change: number;
  resistant_change: number;
  resistant_percentage_change: number;
  hesitant_change: number;
  hesitant_percentage_change: number;
  risk_change: string;
  polarization_change: string;
};

export type PolicySimulationResponse = {
  scenario_name: string;
  population_size: number;
  cycles: number;
  policy: PolicyInput;
  base_scenario: SimulationScenarioSnapshot;
  policy_scenario: SimulationScenarioSnapshot;
  difference: PolicyDifference;
};

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? 'http://127.0.0.1:8000';

export async function runSimulation(payload: SimulationRequest): Promise<SimulationResponse> {
  const response = await fetch(`${API_BASE_URL}/api/simulations/run`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const errorBody = await response.text();
    throw new Error(errorBody || 'Nao foi possivel executar a simulacao.');
  }

  return response.json() as Promise<SimulationResponse>;
}

export async function runPolicySimulation(
  payload: PolicySimulationRequest,
): Promise<PolicySimulationResponse> {
  const response = await fetch(`${API_BASE_URL}/api/policy-simulation/run`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const errorBody = await response.text();
    throw new Error(errorBody || 'Nao foi possivel executar a simulacao de politica.');
  }

  return response.json() as Promise<PolicySimulationResponse>;
}
